HPOS compatibility

alshypilov opened the pull request:
Closes #34 

worked on the next one:
https://docs.google.com/spreadsheets/d/1pRkiBV7oMJhUf-oW95p33zqxJ0fNFhTQE1W1s-zD3pk/edit?usp=sharing

bilfeldt commented on smart-send-logistics/includes/class-ss-shipping-wc-order.php:
```diff
@@ -85,14 +88,17 @@ public function add_meta_box()
             $order_id = $post->ID;
 
             $ss_shipping_method_id = $this->get_smart_send_method_id($order_id);
+            $screen = wc_get_container()->get( CustomOrdersTableController::class )->custom_orders_table_usage_is_enabled()
+                ? wc_get_page_screen_id( 'shop_order' )
+                : 'shop_order';
```
We need to include a link and some description that this is "how to add meta boxes to the legacy order editor screen when legacy orders are in effect, and to the new HPOS-powered editor screen otherwise."

See: https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book#audit-for-order-administration-screen-functions

Do we really need the globals? It seems like we should refactor to a "render_smart_send_metabox" method as well: https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book#audit-for-order-administration-screen-functions

alshypilov replied:
Hi. In my last PR I worked on the following tasks:
I checked in your plugin code and changed next functions:
1)Auditing the code base for direct DB access usage - I checked all the functions and noted which are used in your plugin and which are not in this document (https://docs.google.com/spreadsheets/d/1pRkiBV7oMJhUf-oW95p33zqxJ0fNFhTQE1W1s-zD3pk/edit#gid=0)
2)APIs for getting/setting posts and postmeta - checked
3)Audit for order administration screen functions -started work

Debug log with: Detecting whether - HPOS tables are being used in the store- Do I need to do this check for all the old functions that I will be replacing with new ones for HPOS?

I didn’t work on your project today, I’ll take the whole day tomorrow.

bilfeldt replied:
@AlexShevchenko1985 I moved your comment to the PR itself and answered there https://github.com/smartsendio/woocommerce/pull/41#issuecomment-1874132388

But please see my comment above about the globals and check if this is needed?

bilfeldt reviewed (changes requested):
This looks good 👍 Thanks a lot.

As far as I can see then we need to:
1. Let us for debgging purposes add a debug log entry using [the check for whether or not HPOS is enabled](https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book#detecting-whether-hpos-tables-are-being-used-in-the-store).
2. Add [plugin declaration of HPOS compatability](https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book#declaring-extension-incompatibility)

But can you please fill out below which are done - and let me know if we are missing anything?

- [ ] Debug log with: [Detecting whether HPOS tables are being used in the store](https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book#detecting-whether-hpos-tables-are-being-used-in-the-store)

- [ ] [Auditing the code base for direct DB access usage](https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book#auditing-the-code-base-for-direct-db-access-usage)
- [ ] [APIs for getting/setting posts and postmeta](https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book#apis-for-gettingsetting-posts-and-postmeta)
- [ ] [Audit for order administration screen functions](https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book#audit-for-order-administration-screen-functions)
- [ ] [Declaring extension compatibility](https://github.com/woocommerce/woocommerce/wiki/High-Performance-Order-Storage-Upgrade-Recipe-Book#declaring-extension-incompatibility)

bilfeldt commented:
Note from @alshypilov 

I checked in your plugin code and changed next functions:
- [x] Auditing the code base for direct DB access usage - I checked all the functions and noted which are used in your plugin and which are not in this document (https://docs.google.com/spreadsheets/d/1pRkiBV7oMJhUf-oW95p33zqxJ0fNFhTQE1W1s-zD3pk/edit#gid=0)
- [x] APIs for getting/setting posts and postmeta - checked
- [ ] Audit for order administration screen functions -started work

alshypilov commented:
1)changed the receipt of order ID in the function where we add a metabox to the order admin menu 
 global $woocommerce, $post, $theorder;
 $order_id = $post->ID ?? $theorder->ID;
 Error: Warning: Attempt to read property "ID" on null in /Users/alexshevchenko/Development/smartsendio-new/wp-content/plugins/smart-send-logistics/includes/class-ss-shipping-wc-order.php on line 114...
 
 2)Declaring extension compatibility in class SS_Shipping_WC
 added action add_action('before_woocommerce_init', [$this, 'declaringExtensionCompatibility']);
 public function declaringExtensionCompatibility(): void
        {
            if ( class_exists( \Automattic\WooCommerce\Utilities\FeaturesUtil::class ) ) {
                \Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'custom_order_tables', __FILE__, true );
            }
        }
        
3)Do we really need the globals? It seems like we should refactor to a "render_smart_send_metabox" method as well: 
 -  you mean these globals variables 
<img width="571" alt="image" src="https://github.com/smartsendio/woocommerce/assets/122532662/0257176c-69df-4c0d-b55d-542fc313ac13">
$woocommerce - we don't need it. We need to get the order ID - I looked at how they get it in the woocommerce plugin through the global variable $theorder.
I assume that in older versions you could use $post to get the ID

4)We need to include a link and some description that this is "how to add meta boxes ....  - I don’t fully understand what exactly needs to be added in this fix.
As in our documentation, this metabox will be added in any case, both for HPOS and without HPOS, we have a check.
$screen = wc_get_container()->get( CustomOrdersTableController::class )->custom_orders_table_usage_is_enabled()
                ? wc_get_page_screen_id( 'shop_order' )
                : 'shop_order';
                
5)I think we should just do the debug log for HPOS once. Is there a good place to do this logging? It could be when rendering the admin block or when rendering the settings page? Just one place is fine. -Can you write in more detail what you mean? When activating the plugin, if the HPOS mode is not enabled, do I need to log as in this screenshot?
![image](https://github.com/smartsendio/woocommerce/assets/122532662/77eb7d10-23b5-4281-89c1-437425a3aa55)
As I understand it, this check is needed so that we can use the functionality when HPOS is turned on and if it is turned off, but then we also need to leave all the old functionality and wrap all new functions in this check?
use Automattic\WooCommerce\Utilities\OrderUtil;
if ( OrderUtil::custom_orders_table_usage_is_enabled() ) {
// HPOS usage is enabled.
} else {
// Traditional CPT-based orders are in use.
}
